language/compiler/ir_to_bytecode/syntax/src/ast.rs

```diff
@@ -143,13 +145,30 @@ pub enum StructDefinitionFields {
 #[derive(Debug, Eq, Hash, Ord, PartialEq, PartialOrd, Clone)]
 pub struct FunctionName(String);
 
+/// The kind of a type
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub enum Kind {
+    All,
+    Resource,
+    Unrestricted,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct KindConstraint(String, Kind);
```

Comment:
A newtype like `pub struct TypeParameter(String)` above might be more readable than the raw `String` here?